pkg/cache/sql/informer/listoption_indexer.go

```diff
@@ -191,25 +242,66 @@ func (l *ListOptionIndexer) afterDelete(key string, tx db.TXClient) error {
 	return nil
 }
 
+func (l *ListOptionIndexer) deleteLabels(key string, tx db.TXClient) error {
+	err := tx.StmtExec(tx.Stmt(l.deleteLabelsStmt), key)
+	if err != nil {
+		return &db.QueryError{QueryString: l.deleteLabelsQuery, Err: err}
+	}
+	return nil
+}
+
 // ListByOptions returns objects according to the specified list options and partitions.
 // Specifically:
 //   - an unstructured list of resources belonging to any of the specified partitions
 //   - the total number of resources (returned list might be a subset depending on pagination options in lo)
 //   - a continue token, if there are more pages after the returned one
 //   - an error instead of all of the above if anything went wrong
 func (l *ListOptionIndexer) ListByOptions(ctx context.Context, lo ListOptions, partitions []partition.Partition, namespace string) (*unstructured.UnstructuredList, int, string, error) {
+	queryInfo, err := l.constructQuery(lo, partitions, namespace, db.Sanitize(l.GetName()))
+	if err != nil {
+		return nil, 0, "", err
+	}
+	return l.executeQuery(ctx, queryInfo)
+}
+
+// QueryInfo is a helper-struct that is used to represent the core query and parameters when converting
+// a filter from the UI into a sql query
+type QueryInfo struct {
+	query       string
+	params      []any
+	countQuery  string
+	countParams []any
+	limit       int
+	offset      int
+}
+
+func (l *ListOptionIndexer) constructQuery(lo ListOptions, partitions []partition.Partition, namespace string, dbName string) (*QueryInfo, error) {
+	queryInfo := &QueryInfo{}
+
+	// First, what kind of filtering will we be doing?
 	// 1- Intro: SELECT and JOIN clauses
-	query := fmt.Sprintf(`SELECT o.object, o.objectnonce, o.dekid FROM "%s" o`, db.Sanitize(l.GetName()))
+	// There's a 1:1 correspondence between a base table and its _Fields table
+	// but it's possible that a key has no associated labels, so if we're doing a
+	// non-existence test on labels we need to do a LEFT OUTER JOIN
+	query := fmt.Sprintf(`SELECT o.object, o.objectnonce, o.dekid FROM "%s" o`, dbName)
 	query += "\n  "
-	query += fmt.Sprintf(`JOIN "%s_fields" f ON o.key = f.key`, db.Sanitize(l.GetName()))
+	query += fmt.Sprintf(`JOIN "%s_fields" f ON o.key = f.key`, dbName)
+	if hasLabelFilter(lo.Filters) {
+		query += "\n  "
+		leftOuterPart := ""
+		if hasNegativeExistenceTest(lo.Filters) {
+			leftOuterPart = "LEFT OUTER "
+		}
+		query += fmt.Sprintf(`%sJOIN "%s_labels" lt ON o.key = lt.key`, leftOuterPart, dbName)
```

Comment:
```suggestion
		query += "\n  "
		query += fmt.Sprintf(`LEFT OUTER JOIN "%s_labels" lt ON o.key = lt.key` dbName)
```

Rationale: `LEFT OUTER JOIN`'s results are the same as the default (`INNER`) `JOIN` - only if object rows have no labels `LEFT OUTER JOIN` returns more - filled in with `NULL`s.

Those extra rows will anyway be filtered away by the label conditions in all cases but non-existence, which is what we want.

So if my calculations are right we can simplify this code and let the SQL optimizer do the rest by always using the same `JOIN` type.